Getting started guide improvements - Part 2 (#38288)
rails/rails: guides/source/getting_started.md

controllers and actions.

The line that we have just added says that we are going to match a `GET
/articles` request to `articles#index`. This string passed as the `to` option
represents the _controller_ and _action_ that will be responsible for handling
this request.

Controllers are classes that group together common methods for handling a
particular _resource_. The methods inside controllers are given the name
"actions", as they _act upon_ requests as they come in.

To create a new controller, you will need to run the "controller" generator and
tell it you want a controller called "articles" with an action called "index",
just like this:

```bash
$ bin/rails generate controller articles index
```

Rails will create several files and a route for you.

```bash
create  app/controllers/articles_controller.rb
  route  get 'articles/index'
invoke  erb
create    app/views/articles
create    app/views/articles/index.html.erb
invoke  test_unit
create    test/controllers/articles_controller_test.rb
invoke  helper
create    app/helpers/articles_helper.rb
invoke    test_unit
invoke  assets
invoke    scss
create      app/assets/stylesheets/articles.scss
```

> Most important of these are is of course the controller, located at

Reviewer: `are` should probably be removed.

which is the HTTP status code for successful responses. The rest of the
information on this line is performance metrics.

We have now finished rendering our first view. We created two routes that go to
the same action -- `GET /` and `GET /articles`. That action is the `index`
action in the `ArticlesController`. When we make a request to either of these
paths, the `app/views/articles/index.html.erb` view is rendered, wrapped in the
`app/views/layouts/application.html.erb` layout.

TIP: For more information about routing, refer to [Rails Routing from the Outside In](routing.html).

## Adding Articles

So far, we have seen routes, controllers, actions and views within our Rails
application. All of these are conventional parts of Rails applications and it
is done this way to follow the MVC pattern. The MVC pattern is an application
design pattern which makes it easy to separate the different responsibilities
of applications into easy to reason about pieces.

So with "MVC", you might guess that the "V" stands for "View" and the "C"
stands for controller, but you might have trouble guessing what the "M" stands
for. This next section is all about that "M" part, the _model_.

> ### Generating an Article model

Reviewer: According to the guidelines: 'When writing headings, capitalize all words except for prepositions, conjunctions, internal articles, and forms of the verb "to be"'
https://edgeguides.rubyonrails.org/ruby_on_rails_guides_guidelines.html#headings

So I think `model` should be capitalized.

A model is a class that is used to represent data in our application. In a
plain-Ruby application, you might have a class defined like this:

```ruby
class Article
  attr_reader :title, :body

  def initialize(title:, body:)
    @title = title
    @body = body
  end
end
```

In a Rails application, models are used for more than just representing data.
Models are also used to interact with the application's database. Models use a
part of Rails called Active Record to interact with the database of the
application.

In this section, we're going to use a model to put data into our database and
to pull that data back out, using methods that are provided to us from Active
Record.

To start with, we're going to need to generate a model. We can do that with the
following command:

```bash

> $ bin/rails g model article title:string body:text

Reviewer: I think it's clearer for beginners to avoid shortcuts and use `generate` instead of `g`

Author: I think explaining this earlier on will be better than expanding all `g` to `generate`: 431219185c

migrations in Ruby.

Inside this migration file, there's a `create_table` method that defines how
the `articles` table should be constructed. This method will create a table in
our database that contains an `id` auto-incrementing primary key. That means
that the first record in our table will have an `id` of 1, and the next `id` of
2, and so on.  Rails assumes by default this is the behaviour we want, and so
it does this for us.

Inside the block for `create_table`, we have two fields, `title` and `body`.
These were added to the migration automatically because we put them at the end
of the `rails g model` call:

```bash
$ bin/rails g model article title:string body:text
```

On the last line of the block is `t.timestamps`. This method defines two
additional fields in our table, called `created_at` and `updated_at`. When we
create or update model objects, these fields will be set respectively.

The structure of our table will look like this:

| id | title  | body | created_at | updated_at |
|----|------- |------|------------|------------|
|    | &nbsp; |      |            |            |

To create this table in our application's database, we can run this command:

```bash

> rails db:migrate

Reviewer: The rails command should be written with `bin/rails`:
`$ bin/rails db:migrate`
Adding the $ also makes it look better with the code highlighter.

```

This command will show us output indicating that the table was created:

```plaintext
== [timestamp] CreateArticles: migrating ===================================
-- create_table(:articles)
   -> 0.0018s
== [timestamp] CreateArticles: migrated (0.0018s) ==========================
```

Now that we have a table in our application's database, we can use the model to interact with this table.

To use the model, we'll use a feature of Rails called the _console_. The
console allows us write code like we might in `irb`, but the code of our
application is available there too. Let's launch the console with this command:

```bash

> rails console

Reviewer: The rails command should be written with `bin/rails`:
`$ bin/rails console`

```

This way, we don't need to call `new` and then `save`.

Lastly, models provide a method to find all of their data:

```ruby
irb(main):006:0> articles = Article.all
 #<ActiveRecord::Relation [
   #<Article id: 1, title: "Hello Rails", body: "I am on Rails!", created_at: "2020-01-18 23:47:30", updated_at: "2020-01-18 23:47:30">,
   #<Article id: 2, title: "Post #2", body: "Still riding the Rails!", created_at: "2020-01-18 23:53:45", updated_at: "2020-01-18 23:53:45">]>
```

This method returns an `ActiveRecord::Relation` object, which you can think of
as a super-powered array. This array contains both of the topics that we have
created so far.

TIP: For more about models, read the [Active Record
Basics](https://guides.rubyonrails.org/active_record_basics.html) and [Active
Record Query
Interface](https://guides.rubyonrails.org/active_record_querying.html) guides.

As you can see, models are very helpful classes for interacting with databases
within Rails applications. Models are the final piece of the "MVC" puzzle.
Let's look at how we can go about connecting all these pieces together into a
cohesive whole.

> ### Viewing a list of articles

Reviewer: I think you can capitalize `list` and `articles`

### Setting the Application Home Page

> Now that we have made the route, controller, action and view, let's make a

Reviewer 2: We use the Oxford comma in Rails so this should be:

```suggestion
Now that we have made the route, controller, action, and view, let's make a
```

The third line shows us that Rails is going to process this request by using
the `ArticlesController`, and its `index` action. The _format_ for this request
is HTML. This means that the browser is expecting Rails to return HTML as a
response to this request.

The fourth + fifth lines show that the `articles/index.html.erb` view is being
used in this request. The `layouts/application` mentioned here is located at
`app/views/layouts/application.html.erb`. This file contains the following
content:

```erb
<!DOCTYPE html>
<html>
  <head>
    <title>Blog</title>
    <%= csrf_meta_tags %>
    <%= csp_meta_tag %>

    <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
    <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
  </head>

  <body>
    <%= yield %>
  </body>
</html>
```

This file is the common bits of HTML that will be wrapped around whatever our
views return. The view's output goes right inside that `<body>` tag, where `<%=
yield %>` is.

The next line is:

```
[Webpacker] Everything's up-to-date. Nothing to do
```

Webpacker is a gem that comes with Rails that manages our assets. Right now, we

> don't have any assets and so we can ignore this line, and Webpacker in general.

Reviewer 2: Can you update this to add what assets are? It could be confusing to a newcomer.

The final line is this:

```
Completed 200 OK in 10ms (Views: 7.9ms | ActiveRecord: 0.0ms | Allocations: 4989)
```

It tells us that the request completed successfully. We know it's successful
because we saw it ourselves, but also because this line contains `200 OK`,
which is the HTTP status code for successful responses. The rest of the

> information on this line is performance metrics.

Reviewer 2: ```suggestion
information on this line are performance metrics.
```

We have now finished rendering our first view. We created two routes that go to
the same action -- `GET /` and `GET /articles`. That action is the `index`
action in the `ArticlesController`. When we make a request to either of these
paths, the `app/views/articles/index.html.erb` view is rendered, wrapped in the
`app/views/layouts/application.html.erb` layout.

TIP: For more information about routing, refer to [Rails Routing from the Outside In](routing.html).

## Adding Articles

So far, we have seen routes, controllers, actions and views within our Rails
application. All of these are conventional parts of Rails applications and it
is done this way to follow the MVC pattern. The MVC pattern is an application
design pattern which makes it easy to separate the different responsibilities
of applications into easy to reason about pieces.

> So with "MVC", you might guess that the "V" stands for "View" and the "C"

Reviewer 2: ```suggestion
With "MVC", you might guess that the "V" stands for "View" and the "C"
```

```

This command will show us output indicating that the table was created:

```plaintext
== [timestamp] CreateArticles: migrating ===================================
-- create_table(:articles)
   -> 0.0018s
== [timestamp] CreateArticles: migrated (0.0018s) ==========================
```

Now that we have a table in our application's database, we can use the model to interact with this table.

To use the model, we'll use a feature of Rails called the _console_. The
console allows us write code like we might in `irb`, but the code of our
application is available there too. Let's launch the console with this command:

```bash
$ bin/rails console
```

We can use a shorter version for this too:

```bash
$ bin/rails c
```

Just like we used "g" for "generate", we can use "c" for "console".

When we launch this, we should see an irb prompt:

```plaintext
Loading development environment (Rails 6.0.2.1)
irb(main):001:0>
```

In this prompt, we can use our model to initialize a new `Article` object:

```ruby
irb(main):001:0> article = Article.new(title: "Hello Rails", body: "I am on Rails!")
```

It's important to note that we have just _initialized_ this object. This object

> is not saved to the database at all, it's just available in the console so far.

Reviewer 2: Nitpick word change:

```suggestion
is not saved to the database at all, it's only available in the console so far.
```

irb(main):005:0> Article.create(title: "Post #2", body: "Still riding the Rails!")
```

This way, we don't need to call `new` and then `save`.

Lastly, models provide a method to find all of their data:

```ruby
irb(main):006:0> articles = Article.all
 #<ActiveRecord::Relation [
   #<Article id: 1, title: "Hello Rails", body: "I am on Rails!", created_at: "2020-01-18 23:47:30", updated_at: "2020-01-18 23:47:30">,
   #<Article id: 2, title: "Post #2", body: "Still riding the Rails!", created_at: "2020-01-18 23:53:45", updated_at: "2020-01-18 23:53:45">]>
```

This method returns an `ActiveRecord::Relation` object, which you can think of
as a super-powered array. This array contains both of the topics that we have
created so far.

TIP: For more about models, read the [Active Record
Basics](https://guides.rubyonrails.org/active_record_basics.html) and [Active
Record Query
Interface](https://guides.rubyonrails.org/active_record_querying.html) guides.

As you can see, models are very helpful classes for interacting with databases
within Rails applications. Models are the final piece of the "MVC" puzzle.
Let's look at how we can go about connecting all these pieces together into a

> cohesive whole.

Reviewer 2: I think this last sentence would read better as "Let's look at how we can go about connecting all these pieces together." - especially for those where english is their second language, cohesive might not be a word they know. I think the sentence says what we need to say without "into a cohesive whole"

### Viewing a List of Articles

> Now that you've seen how to create a route, a controller, an action, a view and

Reviewer 2: Oxford comma

```suggestion
Now that you've seen how to create a route, a controller, an action, a view, and
```

a model, let's connect these pieces together.

Let's go back to `app/controllers/articles_controller.rb` now. We're going to
change the `index` action here to use our model.

```ruby
class ArticlesController < ApplicationController
  def index
    @articles = Article.all
  end
end
```

Controller actions are where we assemble all the data that will later be
displayed in the _view_. In this `index` action, we're calling `Article.all`
which will make a query to our database and retrieve all of the articles,
storing them in an instance variable: `@articles`.

We're using an instance variable here for a very good reason: instance
variables are automatically shared from controllers into views. So to use this
`@articles` variable in our view to show all the articles, we can write this
code in `app/views/articles/index.html.erb`:

```erb
<h1>Articles</h1>

<ul>
  <% @articles.each do |article| %>
    <li><%= article.title %></li>
  <% end %>
</ul>
```

We've now changed this file from using just HTML to using HTML and _ERB_. ERB

> is a language that we can use to run Ruby code.

Reviewer 2: I don't want to split hairs here but I think it's confusing calling ERB a language when it ships with Ruby. Ruby calls it a templating system https://ruby-doc.org/stdlib-2.6.5/libdoc/erb/rdoc/ERB.html

```suggestion
is a templating system that we can use to run Ruby code.
```

There's two types of ERB tag beginnings that we're using here: `<%` and `<%=`.
The `<%` tag means to evalulate some Ruby code, while the `<%=` means to
evalulate that code, and then to output the return value from that code.

> In this view, we do not want the output of `articles.each` to show, and so we

Reviewer 2: ```suggestion
In this view, we do not want the output of `@articles.each` to show, and so we
```

and see the list of articles:

![List of articles](images/getting_started/article_list.png)

If we look at the source of the page in our browser
<view-source:http://localhost:3000/>, we'll see this part:

```html
<h1>Articles</h1>

<ul>
    <li>Hello Rails</li>
    <li>Post #2</li>
</ul>
```

This is the HTML that has been output from our view in our Rails application.

Here's what's happened to get to this point:

1. Our browser makes a request: `GET http://localhost:3000`
2. The Rails application receives this request
3. The router sees that the `root` route is configured to route to the `ArticlesController`'s `index` action
4. The `index` action uses the `Article` model to find all the articles
5. Rails automatically renders the `app/views/articles/index.html.erb` view
6. The view contains ERB (Embedded Ruby). This code is evalulated, and plain HTML is returned.
7. The server sends a response containing that plain HTML back to the browser.

Here's a flowchart of the above steps:

![Application flowchart](images/getting_started/application_flowchart.png)

We've now successfully connected all the different parts of our Rails

> application together: the router, the controller, the action, the model and the

Reviewer 2: ```suggestion
application together: the router, the controller, the action, the model, and the
```

Clicking either of these links will take us to the relevant article:

![Single Article View](images/getting_started/single_article_view.png)

Now we have been able to link together the `index` and `show` pages in our
application using a simple `a` tag. What could be simpler than that?

Well, Rails has a method called `link_to` that can make that linking a little
simpler. Let's use this in `app/views/articles/index.html.erb`:

```erb
<h1>Articles</h1>

<ul>
  <% @articles.each do |article| %>
    <li>
      <%= link_to article.title, "/articles/#{article.id}" %>
    </li>
  <% end %>
</ul>
```

There we go, that is now a little bit cleaner. Rails has given us a way to
shorten this code a little. But what you don't know yet is that this line can
be made even simpler.

Rails has a feature called _routing helpers_. These are methods that can be
used to generate route paths like `"/articles/#{article.id}"` programatically.
We'll use one of these to generate the route for our article. To set this up,
let's go back to `config/routes.rb` and change this line:

```ruby
get "/articles/:id", to: "articles#show"
```

To this:

```ruby
get "/articles/:id", to: "articles#show", as: :article

> ```

Reviewer 2: I feel like maybe this is kind of confusing since most of the time you don't type out every route since they're restful. We can say `resources :articles` and get all the routes at once.

I also don't love that we're showing them the wrong way (ie `<%= link_to article.title, "/articles/#{article.id}" %>`) _especially_ since that way _could_ result in a security issue if used incorrectly with another parameter.

Author: I agree that it would be confusing -- to those who have already seen the `resources` syntax elsewhere.

What would you reckon about demonstrating that `resources :articles` _does_ exist, and we _could_ use it, but clearly explaining that we're avoiding using it to demonstrate what that line gives us?

Author: Reading that again, I am not sure I am making 100% sense with that last sentence. Here's a commit that shows the changes I'm suggesting: https://github.com/rails/rails/pull/38288/commits/7121f157885e3e09f96510b9ec6524a1bc8c555f

> We've covered the "CR" part of CRUD. Now let's focus on the "U" part, updating

Reviewer: By replacing the `new` action with a `show` action, the "C" is no longer covered.

Author: This PR does not update this part of the guide, that update will come in a future PR.
